src/librustc_mir/transform/generator.rs

```diff
@@ -1315,3 +1341,134 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
         create_generator_resume_function(tcx, transform, source, body, can_return);
     }
 }
+
+/// Looks for any assignments between locals (e.g., `_4 = _5`) that will both be converted to fields
```

Comment:
Why do we only need to validate assignments where the RHS is a direct usage of a local? Can indirect usages not cause this issue?

Reply:
Good point.

Is MIR such as `*p = *q` or `*p = q` valid when `p` overlaps with `q`? `p = *q` is almost certainly invalid, and should be handled by this check. I'll check for all three versions in a subsequent PR.